No useless index in import paths

timkraut opened the pull request:
As discussed in #394, this PR adds an option `noUselessIndex` to detect (and fix) paths containing unnecessary `index` parts (with or without file extension):

Given the following file structure:
```
my-project
├── foo.js
└── foo
    └── index.js
└── bar
    └── index.js
```

Before:
```
import 'foo/index' // No warning
import 'bar/index' // No warning
```

After:
```
import './foo/index.js' // Warning (fixable to `./foo/`)
import './foo/index' // Warning (fixable to `./foo/`)
import './bar/index' // Warning (fixable to `./bar`)
```

#### Implementation detail:
I've replaced the `sumBy` function with `reduce` to get rid of Lodash (not necessary in this case IMO). A benchmark which compares these 2 alternatives shows no difference between them or a slight advantage for `reduce`: https://www.measurethat.net/Benchmarks/ShowResult/46787

Fixes #394.

timkraut commented:
@ljharb I've updated the PR to use an option `detectUselessIndex` to detect these issue. The rule tries to find ambiguous imports for all extensions provided in:
```js
    'import/resolver': {
      'node': {
        'extensions': [
          // some extensions
        ],
      },
    },
```

If this setting is not set, `['.js']` will be assumed.

Let me know if that looks like a reasonable approach to you. I'm totally open for different ideas.

ljharb commented on src/rules/no-useless-path-segments.js:
```diff
@@ -45,65 +62,99 @@ module.exports = {
+      }
+
+      // Path contains unnecessary /index
+      if (options && options.detectUselessIndex && importPath.endsWith('index')) {
```
should this be checking importPath, or something else? what happens if they’ve specified `index.js`?

timkraut replied:
What do you mean by "should this be checking importPath, or something else?"?

I've added a test for `../index.js`. Importing something like `../index.js` doesn't cause any warnings. The [extensions](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/extensions.md) rule can find those issues. Apparently, the `extension` rule is not able to auto-fix violations. Fixing `../index.js` to be `../` would require 2 steps if both rules are enabled.

ljharb replied:
That means that if a user has `.js` required, they can't get `index.js` auto-removed?

timkraut replied:
Yes, exactly. They would get a warning but the `extension` rule currently is not able to auto-fix those things.

ljharb replied:
It seems like the extensions rule should be unrelated here; any path that could be shortened should be caught by this rule.

timkraut replied:
Ok, I'll add this feature then. Would you consider it part of the `detectUselessIndex` option or add a new option for this?

timkraut replied:
I've updated the PR to find and fix `index.js` as well (depending on the configured file extensions, other file extensions would be found, too). I haven't added a new option for this. Let me know if you prefer an additional option.

ljharb replied:
I think it should be part of `detectUselessIndex`, yes.

ljharb commented on src/rules/no-useless-path-segments.js:
```diff
@@ -45,65 +62,99 @@ module.exports = {
+            path: importPath,
+            proposedPath,
+          },
+          fix: fixer => proposedPath && fixer.replaceText(source, JSON.stringify(proposedPath)),
```
i think that when proposedPath is not present, `fix` shouldn’t be either

timkraut replied:
I've updated the implementation for this case and added a better output handling:
- `fix` is only set if there is a possible fix
- If the auto-fix is not possible because of an ambiguous path, a dedicated message which contains the colliding paths is shown

ljharb commented on src/rules/no-useless-path-segments.js:
```diff
@@ -19,19 +18,37 @@ import docsUrl from '../docsUrl'
+  return pathSegments.reduce((sum, pathSegment) => pathSegment === '..' ? sum + 1 : sum, 0)
+}
+
+function getFileExtensions(settings) {
```
this seems like a useful helper to share in multiple rules (probably one already defined somewhere)

timkraut replied:
Good point! I've found a function to do that - I only had to export it. I've also updated the `README.md` section to clarify how to specify the list of file extensions.

timkraut commented on docs/rules/no-useless-path-segments.md:
````diff
@@ -46,3 +50,24 @@ import ".";
+import "./pages/index"; // should be "./pages" (auto-fixable)
+```
+
+Note: `detectUselessIndex` detects only ambiguous imports for `.js` files if you haven't specified all resolved file extensions. See [Settings: import/extensions](https://github.com/benmosher/eslint-plugin-import#importextensions) for details.
````
Hm... that's difficult to implement. Currently, I loop through all specified extensions for every path which ends in `/index` to make sure there is no ambiguous import (except if the parent directory is `.` or `..` as it's not possible to have ambiguous imports in these cases). If you have an idea how we could do this efficiently for arbitrary file extensions, let me know.

timkraut commented on .gitignore:
```diff
@@ -39,3 +39,6 @@ lib/
 npm-shrinkwrap.json
+
+# IDEs
+.idea/
```
That's a great idea. Never thought about this one. I've updated the PR to exclude this change.

ljharb commented on docs/rules/no-useless-path-segments.md:
````diff
@@ -46,3 +50,24 @@ import ".";
+
+```js
+// in my-project/app.js
+import "./helpers/index"; // should be "./helpers" (not auto-fixable because of helpers.js)
````
this should be autofixable to `./helpers/`

timkraut replied:
Ah. I didn't even know that there is a difference between `./helpers/` --> `./helpers/index.js` and `./helpers` --> `./helpers.js`. 

Just to make sure we don't miss an option, these are all possible import paths:
```
./helpers.js --> ./helpers.js
./helpers --> ./helpers.js
./helpers/ --> ./helpers/index.js
./helpers/index --> ./helpers/index.js
./helpers/index/ --> ./helpers/index/index.js
./helpers/index.js --> ./helpers/index.js
```

If I'm not mistaken, this means that there is no option for ambiguous imports. Either it's an `index.js` file in an directory (in this case, we have a trailing `/`) or it's a file (in this case, we have no trailing `/`). Am I missing something?

ljharb replied:
Yes.

In the presence of both `helpers.js` and `helpers/index.js`:
```sh
./helpers.js → ./helpers.js
./helpers → ./helpers.js (ambiguous)
./helpers/ → ./helpers/index.js
./helpers/index.js → ./helpers/index.js
./helpers/index → ./helpers/index.js
/.helpers/index/ → ./helpers/index/index.js (error, nonexistent here)
```

When there is only `helpers.js`:
```sh
./helpers.js → ./helpers.js
./helpers → ./helpers.js
./helpers/ → ./helpers/index.js (error, nonexistent here)
./helpers/index.js → ./helpers/index.js (error, nonexistent here)
./helpers/index → ./helpers/index.js (error, nonexistent here)
/.helpers/index/ → ./helpers/index/index.js (error, nonexistent here)
```

When there is only `helpers/index.js`:
```sh
./helpers.js → ./helpers.js  (error, nonexistent here)
./helpers → ./helpers/index.js
./helpers/ → ./helpers/index.js
./helpers/index.js → ./helpers/index.js
./helpers/index → ./helpers/index.js
/.helpers/index/ → ./helpers/index/index.js (error, nonexistent here)
```

timkraut replied:
Thank you very much! That helped a lot!

I'll extend the list by the expected result:

### `helpers.js` and `helpers/index.js` exist
Import path (`import '...'`) | Resolves to | Shortest possible path? (Explanation)
-- | -- | --
`./helpers.js` | `./helpers.js` | ✅ (changing the path to `./helpers` would make it ambiguous)
`./helpers` | `./helpers.js` or `./helpers/index.js` (**ambiguous**) | ✅ (path is ambiguous but contains no useless path segments)
`./helpers/` | `./helpers/index.js` | ✅ (changing the path to `./helpers` would make it ambiguous)
`./helpers/index.js` | `./helpers/index.js` | ❌ (change to `./helpers/` possible)
`./helpers/index` | `./helpers/index.js` | ❌ (change to `./helpers/` possible)
`./helpers/index/` | `./helpers/index/index.js` | ~~(Not existent)~~

### Only `helpers.js` exists:
Import path (`import '...'`) | Resolves to | Shortest possible path? (Explanation)
-- | -- | --
`./helpers.js` | `./helpers.js` | ❌ (change to `./helpers` possible)
`./helpers` | `./helpers.js` | ✅ 
`./helpers/` | `./helpers.js` | ❌ (change to `./helpers` possible)
`./helpers/index.js` | `./helpers/index.js` | ~~(Not existent)~~
`./helpers/index` | `./helpers/index.js` | ~~(Not existent)~~
`./helpers/index/` | `./helpers/index/index.js` | ~~(Not existent)~~

### Only `helpers/index.js` exists:
Import path (`import '...'`) | Resolves to | Shortest possible path? (Explanation)
-- | -- | --
`./helpers.js` | `./helpers.js` | ❌ (change to `./helpers` possible)
`./helpers` | `./helpers.js` | ✅ 
`./helpers/` | `./helpers.js` | ❌ (change to `./helpers` possible)
`./helpers/index.js` | `./helpers/index.js` | ❌ (change to `./helpers` possible)
`./helpers/index` | `./helpers/index.js` | ❌ (change to `./helpers` possible)
`./helpers/index/` | `./helpers/index/index.js` | ~~(Not existent)~~

I'll add tests for all these cases and adapt the implementation to that.

ljharb replied:
I've updated your table for two cases, otherwise looks good.

ljharb commented on docs/rules/no-useless-path-segments.md:
````diff
@@ -46,3 +50,24 @@ import ".";
+// in my-project/app.js
+import "./helpers/index"; // should be "./helpers" (not auto-fixable because of helpers.js)
+import "./pages/index"; // should be "./pages" (auto-fixable)
+```
````
can we include an example that has `index.js` as well?

timkraut commented:
I've changed the name of the new option from `detectUselessIndex` to `noUselessIndex` and rebased the branch to the latest `master` commit.

timkraut commented:
I've updated the PR to include tests for `getFileExtensions`. As this function returns a `Set`, I've updated `chai.js` to the latest version to test for that. This resulted in failing tests which I've fixed as well. See https://github.com/chaijs/chai/issues/781 for the breaking changes (in the case of this plugin, I had to rename `.deep` to `.nested`)

timkraut commented:
I'm struggling a little bit with the support for ESLint 2 and 3. I've fixed the issue that `message` is missing which resulted in failing tests. ESLint 4 und 5 seem to understand the `messageId` property. It looks like the failing tests are mostly due to the TypeScript support which isn't working in ESLint 2 and 3 apparently.

I get messages like the following:
```js
errors:
   [ TypeError: traverser.traverse is not a function
         at Object.parseForESLint (/Users/timkraut/Development/eslint-plugin-import/node_modules/typescript-eslint-parser/parser.js:41:15)
         at Object.exports.parse (/Users/timkraut/Development/eslint-plugin-import/node_modules/typescript-eslint-parser/parser.js:79:17)
         at parse (/Users/timkraut/Development/eslint-plugin-import/utils/parse.js:38:17)
         at Function.ExportMap.parse (/Users/timkraut/Development/eslint-plugin-import/src/ExportMap.js:339:15)
         at Function.ExportMap.for (/Users/timkraut/Development/eslint-plugin-import/src/ExportMap.js:323:25)
         at Function.ExportMap.get (/Users/timkraut/Development/eslint-plugin-import/src/ExportMap.js:286:10)
         at Context.<anonymous> (/Users/timkraut/Development/eslint-plugin-import/tests/src/core/getExports.js:332:31)
         at callFn (/Users/timkraut/Development/eslint-plugin-import/node_modules/mocha/lib/runnable.js:348:21)
         at Hook.Runnable.run (/Users/timkraut/Development/eslint-plugin-import/node_modules/mocha/lib/runnable.js:340:7)
         at next (/Users/timkraut/Development/eslint-plugin-import/node_modules/mocha/lib/runner.js:309:10)
         at Immediate.<anonymous> (/Users/timkraut/Development/eslint-plugin-import/node_modules/mocha/lib/runner.js:339:5)
         at processImmediate (timers.js:632:19) ],
```

Do you have an idea how I might fix those?

ljharb commented:
@timkraut i've pulled out your chai change into a separate commit on master (it'll be pushed shortly)

If older eslints don't support messageId, let's not use it yet.

timkraut commented:
Is there anything else I can do to get this merged/released?

timkraut commented:
@ljharb Could you release this, too? Currently the documentation states something which isn't possible with the release from january.

ljharb commented:
Nope, only one person can release this package.
